// @php-wasm/cli-util/index.js
import a from "fs";
import w from "path";
import { toPosixPath as $ } from "@php-wasm/util";
import { XMLParser as F, XMLBuilder as b } from "fast-xml-parser";
import * as x from "jsonc-parser";
const N = "PHPWASMCLI", U = [
  "/dev/",
  "/home/",
  "/internal/",
  "/request/",
  "/proc/"
];
async function W(u, f, l) {
  const d = l === "win32" ? (
    // On Windows, creating a 'dir' symlink can require elevated permissions.
    // In this case, let's make junction points because they function like
    // symlinks and do not require elevated permissions.
    "junction"
  ) : "dir";
  a.symlinkSync(u, f, d);
}
async function q(u) {
  try {
    a.lstatSync(u).isSymbolicLink() && a.unlinkSync(u);
  } catch {
  }
}
function D(u, f) {
  return f.filter((l) => {
    const d = w.resolve(l.hostPath), c = w.join(u, w.sep);
    return (
      // If auto-mounting from the current directory,
      // the entire project directory can be mapped.
      d === u || d.startsWith(c)
    );
  });
}
const E = {
  ignoreAttributes: !1,
  attributeNamePrefix: "",
  preserveOrder: !0,
  cdataPropName: "__cdata",
  commentPropName: "__xmlComment",
  allowBooleanAttributes: !0,
  trimValues: !0
}, I = {
  ignoreAttributes: E.ignoreAttributes,
  attributeNamePrefix: E.attributeNamePrefix,
  preserveOrder: E.preserveOrder,
  cdataPropName: E.cdataPropName,
  commentPropName: E.commentPropName,
  suppressBooleanAttributes: !E.allowBooleanAttributes,
  format: !0,
  indentBy: "	"
}, k = {
  allowEmptyContent: !0,
  allowTrailingComma: !0
};
function X(u, f) {
  var S, C, L, T, _, M;
  const { name: l, host: d, port: c, pathMappings: h, ideKey: t } = f, P = new F(E), v = (() => {
    try {
      return P.parse(u, !0);
    } catch {
      throw new Error("PhpStorm configuration file is not valid XML.");
    }
  })(), y = {
    server: [{}],
    ":@": {
      name: l,
      // NOTE: PhpStorm quirk: Xdebug only works when the full URL (including port)
      // is provided in `host`. The separate `port` field is ignored or misinterpreted,
      // so we rely solely on host: "host:port".
      host: `${d}:${c}`,
      use_path_mappings: "true"
    }
  };
  h && h.length && (y.server[0].path_mappings = h.map(
    (n) => ({
      mapping: [],
      ":@": {
        "local-root": `$PROJECT_DIR$/${$(
          w.relative(f.projectDir, n.hostPath)
        )}`,
        "remote-root": n.vfsPath
      }
    })
  ));
  let p = v == null ? void 0 : v.find(
    (n) => !!(n != null && n.project)
  );
  if (p) {
    const n = (S = p[":@"]) == null ? void 0 : S.version;
    if (n === void 0)
      throw new Error(
        'PhpStorm IDE integration only supports <project version="4"> in workspace.xml, but the <project> configuration has no version number.'
      );
    if (n !== "4")
      throw new Error(
        `PhpStorm IDE integration only supports <project version="4"> in workspace.xml, but we found a <project> configuration with version "${n}".`
      );
  }
  p === void 0 && (p = {
    project: [],
    ":@": { version: "4" }
  }, v.push(p));
  let o = (C = p.project) == null ? void 0 : C.find(
    (n) => {
      var j;
      return !!(n != null && n.component) && ((j = n == null ? void 0 : n[":@"]) == null ? void 0 : j.name) === "PhpServers";
    }
  );
  o === void 0 && (o = {
    component: [],
    ":@": { name: "PhpServers" }
  }, p.project === void 0 && (p.project = []), p.project.push(o));
  let i = (L = o.component) == null ? void 0 : L.find(
    (n) => !!(n != null && n.servers)
  );
  i === void 0 && (i = { servers: [] }, o.component === void 0 && (o.component = []), o.component.push(i));
  const r = (T = i.servers) == null ? void 0 : T.findIndex(
    (n) => {
      var j;
      return !!(n != null && n.server) && ((j = n == null ? void 0 : n[":@"]) == null ? void 0 : j.name) === l;
    }
  );
  (r === void 0 || r < 0) && (i.servers === void 0 && (i.servers = []), i.servers.push(y));
  let e = (_ = p.project) == null ? void 0 : _.find(
    (n) => {
      var j;
      return !!(n != null && n.component) && ((j = n == null ? void 0 : n[":@"]) == null ? void 0 : j.name) === "RunManager";
    }
  );
  if (e === void 0 && (e = {
    component: [],
    ":@": { name: "RunManager" }
  }, p.project === void 0 && (p.project = []), p.project.push(e)), (((M = e.component) == null ? void 0 : M.findIndex(
    (n) => {
      var j;
      return !!(n != null && n.configuration) && ((j = n == null ? void 0 : n[":@"]) == null ? void 0 : j.name) === l;
    }
  )) ?? -1) < 0) {
    const n = {
      configuration: [
        {
          method: [],
          ":@": { v: "2" }
        }
      ],
      ":@": {
        name: l,
        type: "PhpRemoteDebugRunConfigurationType",
        factoryName: "PHP Remote Debug",
        filter_connections: "FILTER",
        server_name: l,
        session_id: t
      }
    };
    e.component === void 0 && (e.component = []), e.component.push(n);
  }
  const m = new b(I).build(v);
  try {
    P.parse(m, !0);
  } catch {
    throw new Error(
      "The resulting PhpStorm configuration file is not valid XML."
    );
  }
  return m;
}
function B(u, f) {
  var p, o, i, r;
  const { pathSkippings: l } = f, d = new F(E), c = (() => {
    try {
      return d.parse(u, !0);
    } catch {
      throw new Error(
        "PhpStorm PHP configuration file is not valid XML."
      );
    }
  })();
  let h = c == null ? void 0 : c.find(
    (e) => !!(e != null && e.project)
  );
  if (h) {
    const e = (p = h[":@"]) == null ? void 0 : p.version;
    if (e === void 0)
      throw new Error(
        'PhpStorm IDE integration only supports <project version="4"> in php.xml, but the <project> configuration has no version number.'
      );
    if (e !== "4")
      throw new Error(
        `PhpStorm IDE integration only supports <project version="4"> in php.xml, but we found a <project> configuration with version "${e}".`
      );
  }
  h === void 0 && (h = {
    project: [],
    ":@": { version: "4" }
  }, c.push(h));
  let t = (o = h.project) == null ? void 0 : o.find(
    (e) => {
      var s;
      return !!(e != null && e.component) && ((s = e == null ? void 0 : e[":@"]) == null ? void 0 : s.name) === "PhpStepFilterConfiguration";
    }
  );
  t === void 0 && (t = {
    component: [],
    ":@": { name: "PhpStepFilterConfiguration" }
  }, h.project === void 0 && (h.project = []), h.project.push(t));
  let P = (i = t.component) == null ? void 0 : i.find(
    (e) => !!(e != null && e.skipped_files)
  );
  if (P === void 0 && (P = { skipped_files: [] }, t.component === void 0 && (t.component = []), t.component.push(P)), l && l.length)
    for (const e of l) {
      const g = `$PROJECT_DIR$${e.endsWith("/") ? e.slice(0, -1) : e}`;
      ((r = P.skipped_files) == null ? void 0 : r.some(
        (S) => {
          var C;
          return !!(S != null && S.skipped_file) && ((C = S == null ? void 0 : S[":@"]) == null ? void 0 : C.file) === g;
        }
      )) || (P.skipped_files === void 0 && (P.skipped_files = []), P.skipped_files.push({
        skipped_file: [],
        ":@": { file: g }
      }));
    }
  const y = new b(I).build(c);
  try {
    d.parse(y, !0);
  } catch {
    throw new Error(
      "The resulting PhpStorm PHP configuration file is not valid XML."
    );
  }
  return y;
}
function O(u, f) {
  var p, o;
  const { name: l, pathMappings: d, pathSkippings: c } = f, h = [];
  let t = u, P = x.parseTree(t, h, k);
  if (P === void 0 || h.length)
    throw new Error("VS Code configuration file is not valid JSON.");
  let v = x.findNodeAtLocation(P, ["configurations"]);
  if (v === void 0 || v.children === void 0) {
    const i = x.modify(t, ["configurations"], [], {});
    t = x.applyEdits(t, i), P = x.parseTree(t, [], k), v = x.findNodeAtLocation(P, [
      "configurations"
    ]);
  }
  const y = (p = v == null ? void 0 : v.children) == null ? void 0 : p.findIndex(
    (i) => {
      var r;
      return ((r = x.findNodeAtLocation(i, ["name"])) == null ? void 0 : r.value) === l;
    }
  );
  if (y === void 0 || y < 0) {
    const i = {
      name: l,
      type: "php",
      request: "launch",
      port: 9003
    };
    d && d.length && (i.pathMappings = d.reduce((s, g) => (s[g.vfsPath] = `\${workspaceFolder}/${$(
      w.relative(f.workspaceDir, g.hostPath)
    )}`, s), {})), c && c.length && (i.skipFiles = c.map(
      (s) => s.endsWith("/") ? `${s}**` : s
    ));
    const r = ((o = v == null ? void 0 : v.children) == null ? void 0 : o.length) || 0, e = x.modify(
      t,
      ["configurations", r],
      i,
      {
        formattingOptions: {
          insertSpaces: !0,
          tabSize: 4,
          eol: `
`
        }
      }
    );
    t = A(t, e);
  }
  return t;
}
async function z({
  name: u,
  ides: f,
  host: l,
  port: d,
  cwd: c,
  mounts: h,
  pathSkippings: t,
  ideKey: P = N
}) {
  const v = h ? D(c, h) : [], y = {};
  if (f.includes("phpstorm")) {
    const p = ".idea/workspace.xml", o = w.join(
      c,
      p
    );
    if (!a.existsSync(o)) {
      if (a.existsSync(w.dirname(o)))
        a.writeFileSync(
          o,
          `<?xml version="1.0" encoding="UTF-8"?>
<project version="4">
</project>`
        );
      else if (f.length == 1)
        throw new Error(
          "PhpStorm IDE integration requested, but no '.idea' directory was found in the current working directory."
        );
    }
    if (a.existsSync(o)) {
      const i = a.readFileSync(o, "utf8"), r = X(i, {
        name: u,
        host: l,
        port: d,
        projectDir: c,
        pathMappings: v,
        ideKey: P
      });
      a.writeFileSync(o, r), y.phpstorm = p;
    }
    if (t && t.length) {
      const i = ".idea/php.xml", r = w.join(
        c,
        i
      );
      if (a.existsSync(r) || a.existsSync(w.dirname(r)) && a.writeFileSync(
        r,
        `<?xml version="1.0" encoding="UTF-8"?>
<project version="4">
</project>`
      ), a.existsSync(r)) {
        const e = a.readFileSync(
          r,
          "utf8"
        ), s = B(e, {
          pathSkippings: t
        });
        a.writeFileSync(r, s), y["phpstorm-php"] = i;
      }
    }
  }
  if (f.includes("vscode")) {
    const p = ".vscode/launch.json", o = w.join(
      c,
      p
    );
    if (!a.existsSync(o)) {
      if (a.existsSync(w.dirname(o)))
        a.writeFileSync(
          o,
          `{
    "configurations": []
}`
        );
      else if (f.length == 1)
        throw new Error(
          "VS Code IDE integration requested, but no '.vscode' directory was found in the current working directory."
        );
    }
    if (a.existsSync(o)) {
      const i = a.readFileSync(o, "utf-8"), r = O(i, {
        name: u,
        workspaceDir: c,
        pathMappings: v,
        pathSkippings: t
      });
      r !== i && (a.writeFileSync(o, r), y.vscode = p);
    }
  }
  return y;
}
async function K(u, f) {
  var h, t, P, v, y;
  const l = w.join(f, ".idea/workspace.xml");
  if (a.existsSync(l)) {
    const p = a.readFileSync(l, "utf8"), o = new F(E), i = (() => {
      try {
        return o.parse(p, !0);
      } catch {
        throw new Error(
          "PhpStorm configuration file is not valid XML."
        );
      }
    })(), r = i.find(
      (m) => !!(m != null && m.project)
    ), e = (h = r == null ? void 0 : r.project) == null ? void 0 : h.find(
      (m) => {
        var S;
        return !!(m != null && m.component) && ((S = m == null ? void 0 : m[":@"]) == null ? void 0 : S.name) === "PhpServers";
      }
    ), s = (t = e == null ? void 0 : e.component) == null ? void 0 : t.find(
      (m) => !!(m != null && m.servers)
    ), g = (P = s == null ? void 0 : s.servers) == null ? void 0 : P.findIndex(
      (m) => {
        var S;
        return !!(m != null && m.server) && ((S = m == null ? void 0 : m[":@"]) == null ? void 0 : S.name) === u;
      }
    );
    if (g !== void 0 && g >= 0) {
      s.servers.splice(g, 1);
      const S = new b(I).build(i);
      try {
        o.parse(S, !0);
      } catch {
        throw new Error(
          "The resulting PhpStorm configuration file is not valid XML."
        );
      }
      S === `<?xml version="1.0" encoding="UTF-8"?>
<project version="4">
	<component name="PhpServers">
		<servers></servers>
	</component>
</project>` ? a.unlinkSync(l) : a.writeFileSync(l, S);
    }
  }
  const d = w.join(f, ".idea/php.xml");
  if (a.existsSync(d)) {
    const p = a.readFileSync(d, "utf8"), o = new F(E), i = (() => {
      try {
        return o.parse(p, !0);
      } catch {
        throw new Error(
          "PhpStorm PHP configuration file is not valid XML."
        );
      }
    })(), r = i.find(
      (s) => !!(s != null && s.project)
    ), e = (v = r == null ? void 0 : r.project) == null ? void 0 : v.findIndex(
      (s) => {
        var g;
        return !!(s != null && s.component) && ((g = s == null ? void 0 : s[":@"]) == null ? void 0 : g.name) === "PhpStepFilterConfiguration";
      }
    );
    if (e !== void 0 && e >= 0) {
      r.project.splice(e, 1);
      const g = new b(I).build(i);
      try {
        o.parse(g, !0);
      } catch {
        throw new Error(
          "The resulting PhpStorm PHP configuration file is not valid XML."
        );
      }
      g === `<?xml version="1.0" encoding="UTF-8"?>
<project version="4">
</project>` ? a.unlinkSync(d) : a.writeFileSync(d, g);
    }
  }
  const c = w.join(f, ".vscode/launch.json");
  if (a.existsSync(c)) {
    const p = [], o = a.readFileSync(c, "utf-8"), i = x.parseTree(o, p, k);
    if (i === void 0 || p.length)
      throw new Error("VS Code configuration file is not valid JSON.");
    const r = x.findNodeAtLocation(i, [
      "configurations"
    ]), e = (y = r == null ? void 0 : r.children) == null ? void 0 : y.findIndex(
      (s) => {
        var g;
        return ((g = x.findNodeAtLocation(s, ["name"])) == null ? void 0 : g.value) === u;
      }
    );
    if (e !== void 0 && e >= 0) {
      const s = x.modify(
        o,
        ["configurations", e],
        void 0,
        {
          formattingOptions: {
            insertSpaces: !0,
            tabSize: 4,
            eol: `
`
          }
        }
      ), g = A(o, s);
      g === `{
    "configurations": []
}` ? a.unlinkSync(c) : a.writeFileSync(c, g);
    }
  }
}
function G({
  cwd: u,
  mounts: f,
  pathSkippings: l
}) {
  return { pathMappings: u && f ? D(u, f) : void 0, pathSkippings: l };
}
function A(u, f) {
  const l = [], d = x.applyEdits(u, f);
  if (l.length = 0, x.parseTree(d, l, k), l.length) {
    const c = l.map((t) => ({
      message: x.printParseErrorCode(t.error),
      offset: t.offset,
      length: t.length,
      fragment: d.slice(
        Math.max(0, t.offset - 20),
        Math.min(d.length, t.offset + t.length + 10)
      )
    })).map(
      (t) => `${t.message} at ${t.offset}:${t.length} (${t.fragment})`
    ), h = f.map(
      (t) => `At ${t.offset}:${t.length} - (${t.content})`
    );
    throw new Error(
      `VS Code configuration file (.vscode/launch.json) is not valid a JSONC after CLI modifications. This is likely a CLI bug. Please report it at https://github.com/WordPress/wordpress-playground/issues and include the contents of your ".vscode/launch.json" file. 

 Applied edits: ${h.join(
        `
`
      )}

 The errors are: ${c.join(`
`)}`
    );
  }
  return d;
}
export {
  N as DEFAULT_IDE_KEY,
  U as DEFAULT_PATH_SKIPPINGS,
  z as addXdebugIDEConfig,
  K as clearXdebugIDEConfig,
  W as createTempDirSymlink,
  G as makeXdebugConfig,
  q as removeTempDirSymlink,
  B as updatePhpStormPHPConfig,
  X as updatePhpStormWorkspaceConfig,
  O as updateVSCodeConfig
};
